examples/image_captioning/model.py

```diff
@@ -0,0 +1,267 @@
+        """Batch of image features to captions."""
+        self.reset(img_feats)
+
+        captions = self.xp.full((img_feats.shape[0], 1), bos, dtype=np.int32)
+        for _ in range(max_caption_length):
+            x = Variable(captions[:, -1])  # Previous word token as input
+            y = self.step(x)
+            pred = y.data.argmax(axis=1).astype(np.int32)
+            captions = self.xp.hstack((captions, pred[:, None]))
+            if (pred == eos).all():
+                break
+        return captions
+
+    def reset(self, img_feats):
+        """Batch of image features to hidden representations.
+
+        Also, reset and then update the internal state of the LSTM.
+        """
+        self.lstm.reset_state()
+        h = self.embed_img(img_feats)
+        h = self.lstm(F.dropout(h, ratio=self.dropout_ratio))
+        return h
+
+    def step(self, x):
+        """Batch of word tokens to word tokens.
+
+        Predict the next set of tokens given previous tokens.
+        """
+        h = self.embed_word(x)
+        h = self.lstm(F.dropout(h, ratio=self.dropout_ratio))
+        h = self.out_word(F.dropout(h, ratio=self.dropout_ratio))
+        return h
+
+
+class NStepLSTMLanguageModel(chainer.Chain):
+    """Recurrent NStepLSTM language model.
+
+    Generate captions given features extracted from images.
+    """
+
+    def __init__(self, vocab_size, hidden_size, dropout_ratio, ignore_label):
+        super(NStepLSTMLanguageModel, self).__init__()
+        with self.init_scope():
+            self.embed_word = L.EmbedID(
+                vocab_size,
+                hidden_size,
+                initialW=initializers.Normal(1.0),
+                ignore_label=ignore_label
+            )
+            self.embed_img = L.Linear(
+                hidden_size,
+                initialW=initializers.Normal(0.01),
+                initial_bias=initializers.Zero()
+            )
+            self.lstm = L.NStepLSTM(1, hidden_size, hidden_size, dropout_ratio)
+            self.decode_caption = L.Linear(
+                hidden_size,
+                vocab_size,
+                initialW=initializers.Normal(0.01),
+                initial_bias=initializers.Zero()
+            )
+
+        self.dropout_ratio = dropout_ratio
+
+    def __call__(self, img_feats, captions):
+        """Batch of image features and image captions to a singe loss.
+
+        Compute the softmax cross-entropy captioning loss in a single pass
+        without iterating over the sequences.
+        """
+        hx, cx, _ = self.reset(img_feats)
+
+        # Extract all inputs and targets for all captions in the batch
+        xs = [c[:-1] for c in captions]
+        ts = [c[1:] for c in captions]
+
+        # Concatenate all input captions and pass them through the model in a
+        # single pass
+        caption_lens = [len(x) for x in xs]
+        caption_sections = np.cumsum(caption_lens[:-1])
+        xs = F.concat(xs, axis=0)
```

Comment:
In my opinion, it seems better that `self.step` includes these 3 lines of concat processing of word indices.
Taking linearly concatenated xs is little unnatural as a function.

Reply:
I'm not too confident about this but is this close to what you meant? https://github.com/chainer/chainer/pull/4076/commits/bdc88f5fc78106a671c6d900a9dce47260178cf1

Reply:
yes!! exactly same as i think!